pennylane/gradients/parameter_shift_cv.py

```diff
@@ -0,0 +1,678 @@
+        operation_descendents = itertools.filterfalse(qml.circuit_graph._is_observable, succ)
+        observable_descendents = filter(qml.circuit_graph._is_observable, succ)
+
+        for BB in operation_descendents:
+            if not BB.supports_heisenberg:
+                # if the descendant gate is non-Gaussian in parameter-shift differentiation
+                # mode, then there must be no observable following it.
+                continue
+
+            B = BB.heisenberg_tr(dev_wires) @ B
+            B_inv = B_inv @ BB.heisenberg_tr(dev_wires, inverse=True)
+
+        Z = B @ Z @ B_inv  # conjugation
+
+        g_tape = tape.copy(copy_operations=True)
+        constants = []
+
+        # transform the descendant observables into their derivatives using Z
+        transformed_obs_idx = []
+
+        for obs in observable_descendents:
+            # get the index of the descendent observable
+            idx = tape.observables.index(obs)
+            transformed_obs_idx.append(idx)
+
+            transformed_obs = _transform_observable(obs, Z, dev_wires)
+
+            A = transformed_obs.parameters[0]
+            constants.append(None)
+
+            # Check if the transformed observable corresponds to a constant term.
+            if len(A.nonzero()[0]) == 1:
+                if A.ndim == 2 and A[0, 0] != 0:
+                    constants[-1] = A[0, 0]
+
+                elif A.ndim == 1 and A[0] != 0:
+                    constants[-1] = A[0]
+
+            g_tape._measurements[idx] = qml.measure.MeasurementProcess(
+                qml.operation.Expectation, _transform_observable(obs, Z, dev_wires)
+            )
+
+        if not any(i is None for i in constants):
+            # Check if *all* transformed observables corresponds to a constant term.
+            # term. If this is the case for all transformed observables on the tape,
+            # then <psi|A|psi> = A<psi|psi> = A,
+            # and we can avoid the device execution.
+            shapes[-1] = 0
+            obs_indices.append(transformed_obs_idx)
+            gradient_values.append(constants)
+            continue
+
+        gradient_tapes.append(g_tape)
+        obs_indices.append(transformed_obs_idx)
+        gradient_values.append(None)
+
+    def processing_fn(results):
+        grads = []
+        start = 0
+
+        if not results:
+            results = [np.zeros([tape.output_dim])]
+
+        interface = qml.math.get_interface(results[0])
+
+        for i, (s, k, ind) in enumerate(zip(shapes, gradient_values, obs_indices)):
+
+            if s == 0:
+                # parameter has zero gradient
+                g = qml.math.zeros_like(results[0], like=interface)
+
+                if k:
+                    g = qml.math.scatter_element_add(g, ind, k, like=interface)
+
+                grads.append(g)
+                continue
+
+            res = results[start : start + s]
+            start = start + s

```

Comment:
Previously in this PR, there seems to have been a bug due to confusion between `res` and `results` in the finite diff logic. How about renaming `res` to `res_for_observable` or so?

Reply:
Good idea, done!